Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+
+        if (type == ReflectionResourceType::Type::ConstantBuffer || type == ReflectionResourceType::Type::StructuredBuffer)
+        {
+            const auto& pElementLayout = pSlangType->getElementTypeLayout();
+            auto& pBufferType = reflectType(pElementLayout, pPath);
+            pType->setStructType(pBufferType);
+        }
+
+        return pType;
+    }
+
+    ReflectionType::SharedPtr reflectStructType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath)
+    {
+        ReflectionStructType::SharedPtr pType = ReflectionStructType::create(getUniformOffset(pPath), pSlangType->getSize());
+        for (uint32_t i = 0; i < pSlangType->getFieldCount(); i++)
+        {
+            ReflectionPath fieldPath;
+            fieldPath.pParent = pPath;
+            fieldPath.pTypeLayout = pSlangType;
+            fieldPath.childIndex = i;
+            fieldPath.pVar = pSlangType->getFieldByIndex(i);
+            ReflectionVar::SharedPtr pVar = reflectVariable(fieldPath.pVar, &fieldPath);
+            pType->addMember(pVar);
+        }
+        return pType;
+    }
+
+    ReflectionType::SharedPtr reflectArrayType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath)
+    {
+        uint32_t arraySize = (uint32_t)pSlangType->getElementCount();
+        uint32_t arrayStride = (uint32_t)pSlangType->getElementStride(SLANG_PARAMETER_CATEGORY_UNIFORM);
+
+        ReflectionPath newPath;
+        newPath.pParent = pPath;
+        newPath.pTypeLayout = pSlangType;
+        newPath.childIndex = 0;
+
+        ReflectionType::SharedPtr pType = reflectType(pSlangType->getElementTypeLayout(), &newPath);
+        ReflectionArrayType::SharedPtr pArrayType = ReflectionArrayType::create(getUniformOffset(pPath), arraySize, arrayStride, pType);
+        return pArrayType;
+    }
+
+    ReflectionType::SharedPtr reflectBasicType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath)
+    {
+        ReflectionBasicType::Type type = getVariableType(pSlangType->getScalarType(), pSlangType->getRowCount(), pSlangType->getColumnCount());
+        ReflectionType::SharedPtr pType = ReflectionBasicType::create(getUniformOffset(pPath), type, false, pSlangType->getSize());
+        return pType;
+    }
+
+    ReflectionType::SharedPtr reflectType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath)
+    {
+        auto kind = pSlangType->getType()->getKind();
+        switch (kind)
+        {
+        case TypeReflection::Kind::Resource:
+        case TypeReflection::Kind::SamplerState:
+        case TypeReflection::Kind::ConstantBuffer:
+        case TypeReflection::Kind::ShaderStorageBuffer:
+        case TypeReflection::Kind::TextureBuffer:
+            return reflectResourceType(pSlangType, pPath);
+        case TypeReflection::Kind::Struct:
+            return reflectStructType(pSlangType, pPath);
+        case TypeReflection::Kind::Array:
+            return reflectArrayType(pSlangType, pPath);
+        default:
```

Comment:
You *might* want to have this explicitly check for scalar/vector/matrix for the `reflectBasicType` case, and have a `default` that leads to an `assert` just so that you can fail fast if/when Slang adds a new kind of type that you don't handle.